incubator/hnc/internal/reconcilers/anchor.go

```diff
@@ -107,123 +110,168 @@ func (r *AnchorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+// shouldDeleteSubns returns true if the namespace still exists and should be deleted as a result of
+// the anchor being deleted. There are a variety of reasons why subnamespaces *shouldn't* be deleted
+// with their anchors (see the body of this method for details); this function also returns false if
+// the subns is already being deleted and we just need to keep waiting for it to disappear.
 //
-// TODO: fix comment post-v0.5
+// This is the only part of the deletion process that needs to access the forest, in order to check
+// the recursive AllowCascadingDeletion setting.
 func (r *AnchorReconciler) shouldDeleteSubns(log logr.Logger, inst *api.SubnamespaceAnchor, nsInst *corev1.Namespace, deletingCRD bool) bool {
 	r.forest.Lock()
 	defer r.forest.Unlock()
 
 	// If the CRD is being deleted, we want to leave the subnamespace alone.
 	if deletingCRD {
+		log.Info("Will not delete subnamespace since the anchor CRD is being deleted")
 		return false
 	}
 
-	// If the anchor isn't in the "ok" state, don't delete the namespace. If it's "conflict," we
-	// *definitely* don't want to delete it; if it's missing (and possibly being created), then there
-	// isn't really a good option but we'll eventually put a condition on the namespace so it's not a
-	// big deal.
-	//
-	// TODO: much of the remaining portion of this function can probably be replaced by this switch
-	// statement. In v0.5, we're playing it safe so I won't modify the rest of the function, but in
-	// v0.6 we should restructure.
+	// The state of the anchor is the most important factor in whether the namespace should be
+	// deleted.
 	switch inst.Status.State {
-	case api.Missing:
-		return false
+	case api.Ok:
+		// The namespace and anchor are bound together, so it's valid to decide the fate of the
+		// namespace here. Firstly, if the namespace is already being deleted, or the anchor has already
+		// been finalized, it means we've already passed step #1 in the deletion process (see onDeleting
+		// for details) so we shouldn't try to perform it again.
+		if !nsInst.DeletionTimestamp.IsZero() {
+			log.V(1).Info("The subnamespace is already being deleted; no need to delete again")
+			return false
+		}
+		if len(inst.ObjectMeta.Finalizers) == 0 {
+			log.V(1).Info("The anchor has already been finalized; do not reconsider deleting the namespace")
+			return false
+		}
+
+		// The subns is ready to be deleted. It's only safe to delete subnamespaces if they're leaves or
+		// if ACD=true on one of their ancestors; the webhooks *should* ensure that this is always true
+		// before we get here, but if something slips by, we'll just leave the old subnamespace alone
+		// with a missing-anchor condition.
+		cns := r.forest.Get(inst.Name)
+		if cns.ChildNames() != nil && !cns.AllowsCascadingDeletion() {
+			log.Info("This subnamespace has descendants and allowCascadingDeletion is disabled; will not delete")
+			return false
+		}
```

Comment:
The new logic is very clear and lgtm! Just that I am a little confused that when would this case happen? CRD is not deleted but the anchor is being deleted when AC is not set and the subns is not a leaf, right? Is this the case when webhook is bypassed?